### ISSUE TITLE
[proposal] task definition best practices

### ISSUE THREAD

**Issue by michaeldauria** [seatgeek/hell#1]
Instead of having the user to populate a `HELL_ENVIRONMENTS`, why not use a capistrano `namespace` instead?

Some suggestions:
- `environments`
- `apps`
- `internal` (hidden from API)

This way, Hell could just loop over all the tasks in the namespace and it can can generate buttons for all environments and apps. There would just have to be documentation that describes which should come first, for me I always have the environment first: `cap <environment> <app> deploy`

**Comment by michaeldauria** [comment]
after an IRL chat about this it seems silly. In its place Hell should have an optional `HELL_APPS` env var that contains a list of the tasks related to an app.

**Comment by josegonzalez** [comment]
The following would be the **nicest** way of doing so:

```
# Possible deploy environments
set :deploy_envs, [ :production, :staging ]

deploy_envs.each do |e|
  task e do
    set :deploy_env, e
    internal.setup_env
    deploy.prepare
  end
end

namespace :environments do
  deploy_envs.each { |e| task e do; end }
end
```

The above, however, is a bit duplicative and sloppy. We need the first stanza, otherwise the cap commands end up looking like the following:

```
cap environments:production deploy
```

That is a real mess, and the same would occur when you do similar for apps:

```
cap app:hell deploy
```

There isn't an way to annotate capistrano commands with extra metadata. I have a few things I can try, but atm I don't have an easy, drop-in solution that does not potentially pollute your top-level namespace.

**Comment by josegonzalez** [comment]
Another proposal is to do the following:

```
# Possible deploy environments
set :deploy_envs, [ :production, :staging ]

deploy_envs.each do |e|
  task e, :tags => [ :hell_env ] do
    set :deploy_env, e
    internal.setup_env
    deploy.prepare
  end
end
```

This change adds a single-line change to your capistrano recipes, and is easy to extend to other features, such as applications, or permissioning etc.

You could also use this to blacklist tasks:

```
task :something_internal, :tags => [ :hell_blacklist ] do
    # you can't perform this through hell
end
```

Thoughts?

**Comment by michaeldauria** [comment]
I like the `tags` approach, its pretty clean and simple to add in. The only issue rolling around in my head is that everything else is configured via environment variables and this would force users to configure Hell in 2 places.

**Comment by josegonzalez** [comment]
What else can we move to the cap file?

**Comment by josegonzalez** [comment]
You can now define an option called `:hell_tags`, which is an array of tags:

```
task :production, :hell_tags => [ :environment ] do
  # custom code here
end
```

Currently only the `:environment` is used for anything, though having an `:application` tag could be cool in the future.

Closed by 70a52033d9bd077f28ec6002903a2019b5fca3cc.
